# PRELIMINARIES

In [1]:
import pandas as pd
import missingno as msn
import os

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.float_format', '{:.2f}'.format)

# DATA LOADING

In [2]:
#Getting a DtypeWarning. Set low_memory=False to get around it
#Low_memory=False instructs pandas to read the entire file before
#inferring data types. It significantly increases memory usage though
#In the path, use trailing dots to go one directory level
orig_df = pd.read_csv('../data/raw/GlobalFindexDatabase2025.csv', low_memory=False)

Check DataFrame info: file is around 28.6 MB

In [3]:
orig_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8564 entries, 0 to 8563
Columns: 437 entries, countrynewwb to con32h_s
dtypes: float64(430), int64(1), object(6)
memory usage: 28.6+ MB


Describe tells us that the indicators are expressed as aggregates (weighted averages)

In [4]:
orig_df.describe()

year     pop_adult  account_t_d  fiaccount_t_d  mobileaccount_t_d  \
count 8564.00       7880.00      8474.00        8382.00            2514.00   
mean  2017.75   38015962.82         0.61           0.58               0.28   
std      4.70  126950906.09         0.28           0.30               0.19   
min   2011.00     228016.00         0.00           0.00               0.00   
25%   2014.00    3739546.00         0.37           0.31               0.13   
50%   2017.00    8517166.00         0.62           0.57               0.24   
75%   2022.00   25903954.00         0.88           0.87               0.41   
max   2024.00 1176678702.00         1.00           1.00               0.94   

       borrow_any_t_d  fin4_d  dig_acc  fin11_2a  fin11a  fin11b  fin11c  \
count         6412.00 1123.00  1375.00    620.00  440.00  512.00  356.00   
mean             0.51    0.35     0.46      0.23    0.18    0.21    0.15   
std              0.15    0.20     0.22      0.12    0.09    0.10    0.07   
min              0.07    0.02     0.03      0.00    0.00    0.00    0.00   
25%              0.40    0.18     0.28      0.14    0.11    0.13    0.12   
50%              0.49    0.32     0.45      0.21    0.17    0.20    0.15   
75%              0.60    0.50     0.63      0.29    0.23    0.27    0.19   
max              0.96    0.86     0.98      0.65    0.44    0.55    0.39   

       fin11f  fin11d  fin11e  fin14a  fin14b  fin14c  fin14d  fin13_1a  \
count  380.00  608.00  740.00  277.00  265.00  336.00  454.00    322.00   
mean     0.17    0.28    0.18    0.20    0.19    0.20    0.31      0.21   
std      0.10    0.16    0.07    0.11    0.10    0.12    0.16      0.08   
min      0.00    0.00    0.00    0.01    0.02    0.03    0.03      0.00   
25%      0.11    0.14    0.12    0.12    0.12    0.12    0.19      0.14   
50%      0.15    0.23    0.16    0.18    0.17    0.16    0.28      0.21   
75%      0.20    0.41    0.22    0.26    0.27    0.25    0.42      0.27   
max      0.63    0.75    0.55    0.51    0.48    0.57    0.75      0.40   

       fin13_1b  fin26a  fin26b  fin27a  fin27b  fin17f  fin17a_17a1_d  \
count    239.00 3170.00 1959.00 1223.00  659.00  442.00        6192.00   
mean       0.15    0.32    0.33    0.24    0.12    0.15           0.28   
std        0.06    0.22    0.23    0.19    0.09    0.10           0.18   
min        0.00    0.00    0.00    0.00    0.00    0.00           0.00   
25%        0.11    0.14    0.14    0.09    0.05    0.07           0.14   
50%        0.14    0.27    0.26    0.18    0.12    0.12           0.23   
75%        0.19    0.48    0.50    0.35    0.18    0.20           0.39   
max        0.31    0.96    0.95    0.92    0.52    0.52           0.88   

       fin17a  fin17b  fin17c  fin22d  fin22e  fin22a_22a1_22g_d  fin22a  \
count 5712.00 1002.00 2932.00 4038.00 1673.00            4956.00 4603.00   
mean     0.27    0.21    0.19    0.19    0.13               0.25    0.25   
std      0.18    0.14    0.10    0.09    0.06               0.17    0.17   
min      0.00    0.00    0.00    0.00    0.01               0.00    0.00   
25%      0.13    0.11    0.11    0.12    0.09               0.13    0.13   
50%      0.22    0.19    0.17    0.17    0.12               0.20    0.20   
75%      0.37    0.28    0.26    0.23    0.16               0.32    0.33   
max      0.88    0.69    0.57    0.68    0.36               0.94    0.94   

       fin22a_1  fin22b  fin22c  fin24sav  fin24fam  fin24work  fin24bor  \
count    344.00 6044.00  958.00   2419.00   2996.00    2493.00    856.00   
mean       0.11    0.28    0.10      0.22      0.36       0.18      0.11   
std        0.10    0.12    0.08      0.15      0.13       0.09      0.06   
min        0.00    0.00    0.00      0.00      0.00       0.00      0.01   
25%        0.03    0.19    0.03      0.12      0.29       0.12      0.08   
50%        0.08    0.27    0.10      0.17      0.36       0.17      0.11   
75%        0.17    0.36    0.15      0.26      0.44       0.2

Check shape

In [5]:
orig_df.shape

(8564, 437)

The shape tells us that there are 8564 rows with 437 columns.

In [6]:
orig_df.dtypes

countrynewwb          object
codewb                object
year                   int64
pop_adult            float64
regionwb24_hi         object
incomegroupwb24       object
group                 object
group2                object
account_t_d          float64
fiaccount_t_d        float64
mobileaccount_t_d    float64
borrow_any_t_d       float64
fin4_d               float64
dig_acc              float64
fin11_2a             float64
fin11a               float64
fin11b               float64
fin11c               float64
fin11f               float64
fin11d               float64
fin11e               float64
fin14a               float64
fin14b               float64
fin14c               float64
fin14d               float64
fin13_1a             float64
fin13_1b             float64
fin26a               float64
fin26b               float64
fin27a               float64
fin27b               float64
fin17f               float64
fin17a_17a1_d        float64
fin17a               float64
fin17b        

Check first 5 rows

In [7]:
orig_df.head(5)

countrynewwb codewb  year   pop_adult  \
0  Afghanistan    AFG  2011 14575546.00   
1      Albania    ALB  2011  2281010.00   
2      Algeria    DZA  2011 26251587.00   
3       Angola    AGO  2011 12779501.00   
4    Argentina    ARG  2011 30685516.00   

                                       regionwb24_hi      incomegroupwb24  \
0                 South Asia (excluding high income)           Low income   
1      Europe & Central Asia (excluding high income)  Upper middle income   
2  Middle East & North Africa (excluding high inc...  Lower middle income   
3         Sub-Saharan Africa (excluding high income)  Lower middle income   
4  Latin America & Caribbean (excluding high income)  Upper middle income   

  group group2  account_t_d  fiaccount_t_d  mobileaccount_t_d  borrow_any_t_d  \
0   all    all         0.09           0.09                NaN             NaN   
1   all    all         0.28           0.28                NaN             NaN   
2   all    all         0.33           0.33                NaN             NaN   
3   all    all         0.39           0.39                NaN             NaN   
4   all    all         0.33           0.33                NaN             NaN   

   fin4_d  dig_acc  fin11_2a  fin11a  fin11b  fin11c  fin11f  fin11d  fin11e  \
0     NaN      NaN       NaN     NaN     NaN     NaN     NaN     NaN     NaN   
1     NaN      NaN       NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2     NaN      NaN       NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3     NaN      NaN       NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4     NaN      NaN       NaN     NaN     NaN     NaN     NaN     NaN     NaN   

   fin14a  fin14b  fin14c  fin14d  fin13_1a  fin13_1b  fin26a  fin26b  fin27a  \
0     NaN     NaN     NaN     NaN       NaN       NaN     NaN     NaN     NaN   
1     NaN     NaN     NaN     NaN       NaN       NaN     NaN     NaN     NaN   
2     NaN     NaN     NaN     NaN       NaN       NaN     NaN     NaN     NaN   
3     NaN     NaN     NaN     NaN       NaN       NaN     NaN     NaN     NaN   
4     NaN     NaN     NaN     NaN       NaN       NaN     NaN     NaN     NaN   

   fin27b  fin17f  fin17a_17a1_d  fin17a  fin17b  fin17c  fin22d  fin22e  \
0     NaN     NaN           0.03    0.03     NaN     NaN     NaN     NaN   
1     NaN     NaN           0.09    0.09     NaN     NaN     NaN     NaN   
2     NaN     NaN           0.04    0.04     NaN     NaN     NaN     NaN   
3     NaN     NaN           0.16    0.16     NaN     NaN     NaN     NaN   
4     NaN     NaN           0.04    0.04     NaN     NaN     NaN     NaN   

   fin22a_22a1_22g_d  fin22a  fin22a_1  fin22b  fin22c  fin24sav  fin24fam  \
0                NaN     NaN       NaN     NaN     NaN       NaN       NaN   
1                NaN     NaN       NaN     NaN     NaN       NaN       NaN   
2                NaN     NaN       NaN     NaN     NaN       NaN       NaN   
3                NaN     NaN       NaN     NaN     NaN       NaN       NaN   
4                NaN     NaN       NaN     NaN     NaN       NaN       NaN   

   fin24work  fin24bor  fin24sell  fin24other  fin24aVD  fin24aSD  fin24aND  \
0        NaN       NaN        NaN         NaN       NaN       NaN       NaN   
1        NaN       NaN        NaN         NaN       NaN       NaN       NaN   
2        NaN       NaN        NaN         NaN       NaN       NaN       NaN   
3        NaN       NaN        NaN         NaN       NaN       NaN       NaN   
4        NaN       NaN        NaN         NaN       NaN       NaN       NaN   

   fin24aSD_ND  fin24aP  fin24aN  fin24sav_SD_ND  fin24fam_SD_ND  \
0          NaN      NaN      NaN             NaN             NaN   
1          NaN      NaN      NaN             NaN             NaN   
2          NaN      NaN      NaN             NaN             NaN   
3          NaN      NaN      NaN             NaN             NaN   
4          NaN      NaN      NaN             NaN             NaN   

   fin24work_SD

The first 5 rows tells us the following:
1. There are 8 dimensions (`countrynewwb`, `codewb`, `year`, `pop_adult`, `regionwb24_hi`, `incomegroupwb24`, `group`, and `group2`)
2. Then the rest of the columns must be the indicators (we will investigate next)

## Indicator Check
According to the Global Findex data, there are "almost 300 indicators on topics such as mobile phone ownership, internet use, digital safety, account ownership, payments, saving, credit, and financial resilience."

In [8]:
#Exclude dimensions
dimensions = ['countrynewwb', 'codewb', 'year', 
              'pop_adult', 'regionwb24_hi', 
              'incomegroupwb24', 'group', 'group2']

columns = orig_df.columns.tolist()
indicators = [indicator for indicator in columns if indicator not in dimensions]
print(f"There are {len(indicators)} indicators in the dataset.")
print("Indicator list:")
print(indicators)

There are 429 indicators in the dataset.
Indicator list:
['account_t_d', 'fiaccount_t_d', 'mobileaccount_t_d', 'borrow_any_t_d', 'fin4_d', 'dig_acc', 'fin11_2a', 'fin11a', 'fin11b', 'fin11c', 'fin11f', 'fin11d', 'fin11e', 'fin14a', 'fin14b', 'fin14c', 'fin14d', 'fin13_1a', 'fin13_1b', 'fin26a', 'fin26b', 'fin27a', 'fin27b', 'fin17f', 'fin17a_17a1_d', 'fin17a', 'fin17b', 'fin17c', 'fin22d', 'fin22e', 'fin22a_22a1_22g_d', 'fin22a', 'fin22a_1', 'fin22b', 'fin22c', 'fin24sav', 'fin24fam', 'fin24work', 'fin24bor', 'fin24sell', 'fin24other', 'fin24aVD', 'fin24aSD', 'fin24aND', 'fin24aSD_ND', 'fin24aP', 'fin24aN', 'fin24sav_SD_ND', 'fin24fam_SD_ND', 'fin24work_SD_ND', 'fin24bor_SD_ND', 'fin24sell_SD_ND', 'fin24other_SD_ND', 'fin24sav_VD', 'fin24fam_VD', 'fin24work_VD', 'fin24bor_VD', 'fin24sell_VD', 'fin24other_VD', 'fh1', 'fin28', 'fh2', 'fin29', 'fin31a_31b', 'fin30', 'fin31a', 'fin31b', 'fin31d', 'fin32_33_34a', 'fin32_33_34b', 'fin32_33_34c', 'fin32_33_34d', 'fin32_33_acc', 'fin32_33', 'f

There are 429 indicators in the list. We have gone through the Glossary of these indicators and have found that some of them are the same indicator but with different expressions. For example, `con26d`, `con26w`, and `con26m` are internet use, aggregated daily, weekly, and monthly respectively. Therefore we think that there might be 300 **unique** indicators in the list, while the rest are different aggregations of the data.

## Country Check

In [9]:
#Check unique countries
countries = orig_df.codewb.unique()
print(f"There are {len(countries)} countries in the dataset.")
print("Country list:")
print(countries)

There are 174 countries in the dataset.
Country list:
['AFG' 'ALB' 'DZA' 'AGO' 'ARG' 'ARM' 'AUS' 'AUT' 'AZE' 'BHR' 'BGD' 'BLR'
 'BEL' 'BEN' 'BOL' 'BIH' 'BWA' 'BRA' 'BGR' 'BFA' 'BDI' 'KHM' 'CMR' 'CAN'
 'CAF' 'TCD' 'CHL' 'CHN' 'COL' 'COM' 'COD' 'COG' 'CRI' 'HRV' 'CYP' 'CZE'
 'DNK' 'DJI' 'DOM' 'ECU' 'EGY' 'SLV' 'EST' 'SWZ' 'FIN' 'FRA' 'GAB' 'GEO'
 'DEU' 'GHA' 'GRC' 'GTM' 'GIN' 'HTI' 'HND' 'HKG' 'HUN' 'IND' 'IDN' 'IRN'
 'IRQ' 'IRL' 'ISR' 'ITA' 'JAM' 'JPN' 'JOR' 'KAZ' 'KEN' 'KOR' 'XKX' 'KWT'
 'KGZ' 'LAO' 'LVA' 'LBN' 'LSO' 'LBR' 'LTU' 'LUX' 'MDG' 'MWI' 'MYS' 'MLI'
 'MLT' 'MRT' 'MUS' 'MEX' 'MDA' 'MNG' 'MNE' 'NPL' 'NLD' 'NZL' 'NIC' 'NER'
 'NGA' 'MKD' 'OMN' 'PAK' 'PAN' 'PRY' 'PER' 'PHL' 'POL' 'PRT' 'QAT' 'ROU'
 'RUS' 'RWA' 'SAU' 'SEN' 'SRB' 'SLE' 'SGP' 'SVK' 'SVN' 'ZAF' 'ESP' 'LKA'
 'SDN' 'SWE' 'SYR' 'TWN' 'TJK' 'TZA' 'THA' 'TGO' 'TTO' 'TUR' 'TKM' 'UGA'
 'UKR' 'ARE' 'GBR' 'USA' 'URY' 'UZB' 'VEN' 'VNM' 'PSE' 'YEM' 'ZMB' 'ZWE'
 'BLZ' 'BTN' 'CIV' 'ETH' 'MMR' 'NAM' 'NOR' 'PRI' 'SOM' 'CHE' 'TUN' 'GM

For this analysis, we are interested in the ASEAN 5, including Singapore and the Global average:
1. `PHL`: Philippines
2. `IDN`: Indonesia
3. `THA`: Thailand
4. `VNM`: Vietnam
5. `MYS`: Malaysia
6. `SGP`: Singapore
7. `WLD`: World

## Year Check

In [10]:
#Check years
years = orig_df.year.unique()
print(f"There are {len(years)} countries in the dataset.")
print("Year list:")
print(years)

There are 6 countries in the dataset.
Year list:
[2011 2014 2017 2021 2022 2024]


Confirming that the study spans 13 years. We are only interested in the last 10 years, so we will exclude 2011 in our analysis.

# DATA PREPARATION
In this section, we'll prepare the dataset we're going to use for our analysis.

1. We will keep only ASEAN 5, Singapore, World `country` values.
2. We will remove 2011 in the `year` values
3. We will keep the following indicators (please see the next section)

## SECTION HERE ON INDICATORS TO USE

In [11]:
df = orig_df.copy()

In [12]:
asean_5 = ['PHL', 'IDN', 'THA', 'VNM', 'MYS', 'SGP', 'WLD']

In [13]:
#Filter for Asean 5
df = df[df['codewb'].isin(asean_5)]

In [14]:
#Remove 2011
df = df[~df['year'].isin([2011])]

In [15]:
df.shape

(319, 437)

We're down to 394 rows!

In [16]:
#Recheck countries
countries = df.codewb.unique()
print(f"There are {len(countries)} countries in the dataset.")
print("Country list:")
print(countries)

There are 7 countries in the dataset.
Country list:
['IDN' 'MYS' 'PHL' 'SGP' 'THA' 'VNM' 'WLD']


In [17]:
#Recheck years
years = df.year.unique()
print(f"There are {len(years)} countries in the dataset.")
print("Year list:")
print(years)

There are 5 countries in the dataset.
Year list:
[2014 2017 2021 2022 2024]


In [18]:
#Check info
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 319 entries, 202 to 7936
Columns: 437 entries, countrynewwb to con32h_s
dtypes: float64(430), int64(1), object(6)
memory usage: 1.1+ MB
None


## Saving processed data for analysis

In [19]:
file_path = '../data/processed/global_findex_for_analysis.csv'

if not os.path.exists(file_path):
    df.to_csv(file_path, index=False)
    print("File saved successfully!")
else:
    print("File already exists. Skipping save.")

File already exists. Skipping save.


We have now filtered the data for what we need for our analysis. We will then investigate the missingness of the data and see how we must handle them.

# HANDLING MISSINGNESS
(Draft still but):
1. "If the variable is not gonna be needed, it does not matter whether the particular item is missing or not and should be excluded in the subset of dataframe before applying listwise deletion." - so by selecting the variables that we need, we already have handled an important part here.
2. Since this is a survey, I think the data is **MCAR**. It is most likely that the missing data is just because the variable is not being measure at that year. My hunch is that most of these data are in missing years (will have to confirm).